### ISSUE TITLE
`anchor build` freezes by nesting files with the same name

### ISSUE THREAD

**Issue by Tristyn** [coral-xyz/anchor#2355]
`anchor build` will never complete and IDL will not be generated if the source contains a module tree like this:
```
lib.rs
freeze.rs
freeze/freeze.rs
```

I would have expected the build to complete or an error about module names to be printed.

Interestingly, if freeze.rs is overwritten while the build is ongoing, the build will complete.

Reproduction: https://github.com/Tristyn/anchor-build-hangs

versions:
avm 0.24.2
anchor-cli 0.25.0
solana-cli 1.13.4 (src:3b81ee9c; feat:4011803773)
stable-x86_64-apple-darwin (default)
rustc 1.66.0 (69f9c33d7 2022-12-12)

Build log:

```
    Finished release [optimized] target(s) in 1.40s
cargo-build-bpf child: /Users/tristyn/.local/share/solana/install/releases/1.13.4/solana-release/bin/sdk/bpf/scripts/strip.sh /Users/tristyn/repos/build-hangs/target/bpfel-unknown-unknown/release/build_hangs.so /Users/tristyn/repos/build-hangs/target/deploy/build_hangs.so
cargo-build-bpf child: /Users/tristyn/.local/share/solana/install/releases/1.13.4/solana-release/bin/sdk/bpf/dependencies/bpf-tools/llvm/bin/llvm-readelf --dyn-symbols /Users/tristyn/repos/build-hangs/target/deploy/build_hangs.so

To deploy this program:
  $ solana program deploy /Users/tristyn/repos/build-hangs/target/deploy/build_hangs.so
The program address will default to this keypair (override with --program-id):
  /Users/tristyn/repos/build-hangs/target/deploy/build_hangs-keypair.json
# This part will never complete
```

**Comment by Henry-E** [comment]
do you have any clues as to which part of anchor is freezing? It seems like the kind of thing that might affect the IDL generation code rather than any of the macros since you're not actually using anything from `freeze` only exporting and passing around modules.

I would look through the IDL files scraping code maybe, I haven't had a chance to look at it but i don't know else could be causing issues.

**Comment by Henry-E** [comment]
Actually here's a possible similar issue?  https://github.com/coral-xyz/anchor/issues/899

**Comment by Tristyn** [comment]
> It seems like the kind of thing that might affect the IDL generation code

Yes it freezes before it can output an IDL, and still freezes with the --skip-lint flag

**Comment by Henry-E** [comment]
If the issue is that there can't be more than one module with the same name because of used btreemap used by the IDL parser, then I don't think there's much that can be done except adding an error if the IDL code tries to add a second module of the same name.

**Comment by wowinter13** [comment]
@Henry-E @acheroncrypto what do you think about closing this one? As it was fixed @ https://github.com/coral-xyz/anchor/pull/2824

**Comment by acheroncrypto** [comment]
Sure, we can close. However, note that we're still using the parser during the IDL build step, so the issue might *not* have been fixed fully. I do plan to get rid of that part soon though. It's very hacky and also requires us to use the `nightly` compiler.
